File: .github/scripts/bulk-issue-report.py
import requests
import os
from datetime import datetime, timezone

# Di Actions, GITHUB_TOKEN otomatis tersedia dari environment runner
# Tidak perlu set manual seperti saat jalankan lokal
token = os.environ["GITHUB_TOKEN"]
owner = os.environ["GITHUB_REPOSITORY_OWNER"]
repo = os.environ["GITHUB_REPOSITORY"].split("/")[1]

# Header HTTP untuk autentikasi dan versi API
headers = {
    "Authorization": f"Bearer {token}",
    "Accept": "application/vnd.github+json",  # format response GitHub API v3
    "X-GitHub-Api-Version": "2022-11-28",     # versi API yang kita minta
}

print(f"\n📋 Bulk Issue Report (Python requests - Actions)")
print(f"Repo: {owner}/{repo}")
print(f"================================\n")

# Pagination manual — loop sampai tidak ada data lagi
semua_issues = []
page = 1

while True:
    # Request ke endpoint issues GitHub
    response = requests.get(
        f"https://api.github.com/repos/{owner}/{repo}/issues",
        headers=headers,
        params={
            "state": "open",
            "per_page": 100,
            "page": page,
        }
    )

    # Hentikan jika request gagal
    response.raise_for_status()

    data = response.json()

    # Halaman kosong = semua data sudah diambil
    if not data:
        break

    semua_issues.extend(data)
    page += 1

print(f"Total issue terbuka: {len(semua_issues)}\n")

sekarang = datetime.now(timezone.utc)

for issue in semua_issues:
    # Parse tanggal dari format ISO 8601 GitHub ke objek datetime Python
    dibuat = datetime.fromisoformat(issue["created_at"].replace("Z", "+00:00"))
    usia_hari = (sekarang - dibuat).days

    labels = issue["labels"]
    nama_labels = ", ".join(l["name"] for l in labels) if labels else "tanpa label"

    print(f"#{issue['number']} — {issue['title']}")
    print(f"  Usia    : {usia_hari} hari")
    print(f"  Label   : {nama_labels}")
    print(f"  Pembuat : {issue['user']['login']}")
    print()

# Statistik ringkasan untuk log Actions
issues_dengan_label = sum(1 for i in semua_issues if i["labels"])
issues_tanpa_label = sum(1 for i in semua_issues if not i["labels"])

print(f"================================")
print(f"Statistik:")
print(f"  Punya label    : {issues_dengan_label} issue")
print(f"  Tanpa label    : {issues_tanpa_label} issue")
print(f"================================\n")

File: .github/scripts/bulk-issue-report.js
import { Octokit } from "@octokit/action";

// Untuk Actions: pakai @octokit/action yang otomatis ambil GITHUB_TOKEN
// Berbeda dengan bulk-issue-report-lokal.js yang pakai @octokit/rest
const octokit = new Octokit();

// Di Actions, variabel ini otomatis tersedia dari environment runner
const owner = process.env.GITHUB_REPOSITORY_OWNER;
const repo = process.env.GITHUB_REPOSITORY.split("/")[1];

console.log(`\n📋 Bulk Issue Report (Actions)`);
console.log(`Repo: ${owner}/${repo}`);
console.log(`================================\n`);

// Ambil SEMUA issue terbuka menggunakan paginate
// paginate otomatis handle semua halaman tanpa perlu urus pagination manual
const issues = await octokit.paginate(octokit.rest.issues.listForRepo, {
  owner,
  repo,
  state: "open",    // hanya issue yang masih terbuka
  per_page: 100,    // maksimal per halaman untuk kurangi jumlah request
});

console.log(`Total issue terbuka: ${issues.length}\n`);

// Hitung tanggal hari ini untuk kalkulasi usia issue
const sekarang = new Date();

// Loop setiap issue dan tampilkan informasinya di log Actions
for (const issue of issues) {
  // Hitung usia issue dalam hari sejak dibuat
  const dibuat = new Date(issue.created_at);
  const selisihMs = sekarang - dibuat;
  const usiaHari = Math.floor(selisihMs / (1000 * 60 * 60 * 24));

  // Ambil nama semua label yang terpasang, atau "tanpa label" jika kosong
  const labels = issue.labels.length > 0
    ? issue.labels.map(l => l.name).join(", ")
    : "tanpa label";

  // Tampilkan ringkasan tiap issue di log Actions
  console.log(`#${issue.number} — ${issue.title}`);
  console.log(`  Usia    : ${usiaHari} hari`);
  console.log(`  Label   : ${labels}`);
  console.log(`  Pembuat : ${issue.user.login}`);
  console.log();
}

// Ringkasan statistik di akhir laporan
const issuesDenganLabel = issues.filter(i => i.labels.length > 0).length;
const issuesTanpaLabel = issues.filter(i => i.labels.length === 0).length;

console.log(`================================`);
console.log(`Statistik:`);
console.log(`  Punya label    : ${issuesDenganLabel} issue`);
console.log(`  Tanpa label    : ${issuesTanpaLabel} issue`);
console.log(`================================\n`);
